core/policy/policy_repository.go

```diff
@@ -0,0 +1,168 @@
+func NewPolicyRepository(client *requests.HTTPClient, policyURL string, interval time.Duration) *PolicyRepository {
+	return &PolicyRepository{
+		client:        client,
+		policyURL:     policyURL,
+		policyRules:   make(map[market.AccessPolicy]market.AccessPolicyRuleSet),
+		fetchInterval: interval,
+	}
+}
+
+// Start begins fetching proposals to repository
+func (pr *PolicyRepository) Start() {
+	pr.fetchShutdown = make(chan bool, 1)
+	go pr.fetchLoop()
+}
+
+// Stop ends fetching proposals to repository
+func (pr *PolicyRepository) Stop() {
+	pr.fetchShutdown <- true
+}
+
+// Policy converts given value to valid policy rule
+func (pr *PolicyRepository) Policy(policyID string) market.AccessPolicy {
+	return market.AccessPolicy{
+		ID:     policyID,
+		Source: fmt.Sprintf("%v%v", pr.policyURL, policyID),
+	}
+}
+
+// Policies converts given values to list of valid policies
+func (pr *PolicyRepository) Policies(policyIDs []string) []market.AccessPolicy {
+	policies := make([]market.AccessPolicy, len(policyIDs))
+	for i, policyID := range policyIDs {
+		policies[i] = pr.Policy(policyID)
+	}
+	return policies
+}
+
+// AddPolicies adds given policy to repository. Also syncs policy rules from TrustOracle
+func (pr *PolicyRepository) AddPolicies(policies []market.AccessPolicy) error {
+	pr.policyLock.Lock()
+	policyRulesNew := pr.policyRules
+	pr.policyLock.Unlock()
+
+	for _, policy := range policies {
+		policyRules, err := pr.fetchPolicyRules(policy)
+		if err != nil {
+			return errors.Wrap(err, "initial fetch failed")
+		}
+
+		policyRulesNew[policy] = policyRules
+	}
+
+	pr.policyLock.Lock()
+	pr.policyRules = policyRulesNew
+	pr.policyLock.Unlock()
+
+	return nil
+}
+
+// RulesForPolicy gives rules of given policy
+func (pr *PolicyRepository) RulesForPolicy(policy market.AccessPolicy) (market.AccessPolicyRuleSet, error) {
+	pr.policyLock.Lock()
+	defer pr.policyLock.Unlock()
+
+	policyRules, exist := pr.policyRules[policy]
+	if !exist {
+		return policyRules, fmt.Errorf("unknown policy: %s", policy)
+	}
+	return policyRules, nil
+}
+
+// RulesForPolicies gives list of rules of given policies
+func (pr *PolicyRepository) RulesForPolicies(policies []market.AccessPolicy) ([]market.AccessPolicyRuleSet, error) {
+	pr.policyLock.Lock()
+	defer pr.policyLock.Unlock()
+
+	policiesRules := make([]market.AccessPolicyRuleSet, len(policies))
+	for i, policy := range policies {
+		policyRules, exist := pr.policyRules[policy]
+		if !exist {
+			return policiesRules, fmt.Errorf("unknown policy: %s", policy)
+		}
+		policiesRules[i] = policyRules
+	}
+	return policiesRules, nil
+}
+
+func (pr *PolicyRepository) fetchPolicyRules(policy market.AccessPolicy) (market.AccessPolicyRuleSet, error) {
+	var policyRules market.AccessPolicyRuleSet
+
+	req, err := requests.NewGetRequest(policy.Source, "", nil)
+	if err != nil {
+		return policyRules, errors.Wrap(err, "failed to create policy request")
+	}
+
+	err = pr.client.DoRequestAndParseResponse(req, &policy)
```

Comment:
Shouldn't it be `pr.client.DoRequestAndParseResponse(req, &policyRules)` ? I mean response is parsed into policy not policyRules

Reply:
Good point, thanks